### ISSUE TITLE
BUG: max, argmax NaN handling with MPS backend vs CPU backend

### ISSUE THREAD

**Issue by tylerjereddy** [pytorch/pytorch#130295]
I don't see any related MPS `max` reproducers quite this simple on the issue tracker, so figured this might help. Working on an M3 Max, Python `3.12` with `torch` `2.3.1`, I saw the `max` discrepancy vs. CPU below. This came up while I was investigating array API conformance in SciPy. I wonder if other more complex neural network MPS issues like https://github.com/pytorch/pytorch/issues/122030 might ultimately reduce to problems similar to this given that `argmax` is also similarly affected and obviously heavily used.

I was initially thinking this behavior might be "ok" because of https://github.com/pytorch/pytorch/issues/116567#issuecomment-1917924375, where it is clarified that NaN sort-order is unspecified in `torch`. However, the array API specifies a requirement for NaN propagation from `max` (https://data-apis.org/array-api/latest/API_specification/generated/array_api.max.html#max) and the `cpu` device is already conformant.

Running:

```python
import torch 

print(torch.__version__)
for device_name in ["cpu", "mps"]:
    tensor = torch.tensor([1.0, float("nan"), 3.0], device=torch.device(device_name))
    max = torch.max(tensor)
    print(f"{device_name=}, {max=}")
```

Produces:

```
2.3.1
device_name='cpu', max=tensor(nan)
device_name='mps', max=tensor(3., device='mps:0')
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by jhavukainen** [comment]
Yeah I see the problem here, we should conform to the NaN propagation convention in the reduction ops here. We'll need to change the MPS min max calls to use the `reductionMaximumPropagateNaNWithTensor: axes: name:` API instead to enable those for max and min. 

Unfortunately with argmax and argmin I'm not seeing a similar NaN propagating API being available, I'll need to raise an issue with the MPSGraph team to see if they have a workaround or if they can add a proper API for this. In case there is no workaround I'll see about adding a torch check to raise error of improper behavior if NaNs are detected in the inputs while waiting for a solution to be added to the graph side.

**Comment by jhavukainen** [comment]
Thanks for reporting this @tylerjereddy. I added the easy part of the fix and a regression test based on your snippet here https://github.com/pytorch/pytorch/pull/130445

For the argmin/argmax case I'll need to wait for some internal feedback first.

**Comment by skotapati** [comment]
The associated PR has been merged, and I can confirm that this no longer repros with the latest pytorch nightly. Closing the issue

**Comment by malfet** [comment]
Updated title to reflect that argmax/argmin NaN handling is wrong for MPS backend

**Comment by malfet** [comment]
MPS argmin/argmax still does not handle NaNs correctly, but as I'm working on torch.compile perf optimizaiton, may be I'll be able to write shaders for those as well... My simple reproducer
```
% python3 -c "import torch;x=torch.rand(32, device='mps'); x[5]=torch.nan;print(x.argmin(), x.argmax())"
tensor(13, device='mps:0') tensor(3, device='mps:0')
```

**Comment by malfet** [comment]
And my shader that computes argmax for simdgroup correctly
```metal
kernel void argmax(device float* x, device int* y, uint thread_index [[thread_position_in_grid]]) {
   const auto rc = metal::simd_max(x[thread_index]);
   const auto vote = metal::simd_ballot(x[thread_index] == rc || metal::isnan(x[thread_index]));
   y[thread_index] = metal::ctz(static_cast<uint>(static_cast<ulong>(vote)));
}
```
